validator/main.go

```diff
@@ -0,0 +1,241 @@
+		log.Fatalf("%#v", err)
+	}
+
+	v := decodeFile(configFile)
+	expandedConfig := replaceVariables(v)
+
+	if err = schema.Validate(expandedConfig); err != nil {
+		if ve, ok := err.(*jsonschema.ValidationError); ok {
+			log.Fatalf("%#v", ve)
+		} else {
+			log.Fatalf("%#v", err)
+		}
+	}
+
+	return expandedConfig
+}
+
+func decodeFile(file string) interface{} {
+	data, err := os.ReadFile(file)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	ext := filepath.Ext(file)
+	if ext == ".yaml" || ext == ".yml" {
+		return decodeYAML(file)
+	}
+
+	var v interface{}
+	if err := json.Unmarshal(data, &v); err != nil {
+		log.Fatalf("Invalid json file %s: %#v", file, err)
+	}
+
+	return v
+}
+
+func decodeYAML(file string) interface{} {
+	var v interface{}
+
+	body, err := os.ReadFile(file)
+	if err != nil {
+		log.Fatalf("Failed to read configuration file %s: %v", file, err)
+	}
+
+	reader := bytes.NewReader(body)
+	dec := yaml.NewDecoder(reader)
+
+	if err := dec.Decode(&v); err != nil {
+		log.Fatalf("Invalid yaml file %s: %v", file, err)
+	}
+
+	return v
+}
+
+func jsonToFile(j interface{}, outFile string) {
+	ext := filepath.Ext(outFile)
+	if ext == ".yaml" || ext == ".yml" {
+		yamlString, err := yaml.Marshal(j)
+		err = os.WriteFile(outFile, yamlString, 0644)
+		if err != nil {
+			log.Fatalf("Unable to write output file: %v", err)
+		}
+
+		err = os.WriteFile(outFile, yamlString, 0644)
+		if err != nil {
+			log.Fatalf("Unable to write output file: %v", err)
+		}
+	} else if ext == ".json" {
+		jsonString, err := json.MarshalIndent(j, "", "  ")
+		if err != nil {
+			log.Fatalf("Unable to convert to json: %v", err)
+		}
+
+		err = os.WriteFile(outFile, jsonString, 0644)
+		if err != nil {
+			log.Fatalf("Unable to write output file: %v", err)
+		}
+	} else {
+		log.Fatalf("Unknown extension on output file %v", outFile)
+	}
+}
+
+func replaceVariables(c interface{}) interface{} {
+	expandedConfig := make(map[string]any)
+	m, _ := c.(map[string]any)
+	for k := range m {
+		val := expandValues(m[k])
+		expandedConfig[k] = val
+	}
+
+	return expandedConfig
+}
+
+func expandValues(value any) any {
+	switch v := value.(type) {
+	case string:
+		if !strings.Contains(v, "${") || !strings.Contains(v, "}") {
+			return v
+		}
+
+		return expandString(v)
+	case []any:
+		l := []any{}
+		for _, e := range v {
+			newElement := expandValues(e)
+			l = append(l, newElement)
+		}
+		return l
+	case map[string]any:
+		newMap := make(map[string]any)
+
+		for k, v := range v {
+			updated := expandValues(v)
+			newMap[k] = updated
+		}
+
+		return newMap
+	}
+
+	return value
+}
+
+// Replace environment variables ${EXAMPLE} with their value and continue to
+// try replacing variables until there are no more, meaning ${EXAMPLE} could
+// contain another variable ${ANOTHER_VARIABLE}. But stop after 100 iterations
+// to prevent an infinite loop.
+// This does not use os.ExpandVars in order to support defaults like ${VAR:default}
+func expandString(s string) string {
```

Comment:
Did not test, but just from reading this comment, the behavior is violating the spec:

* https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/configuration/file-configuration.md#environment-variable-substitution

> It MUST NOT be possible to inject environment variable by environment variables. For example, references to DO_NOT_REPLACE_ME environment variable below.

Reply:
Oh, hm, I could have swore I included this because it was in the spec that you can do this. Thanks for calling it out, I'll remove the extra for-loop.

Reply:
Fixed.